eclair-core/src/main/scala/fr/acinq/eclair/channel/TxPublisher.scala

```diff
@@ -0,0 +1,413 @@
+              val cltvTimeout = Scripts.cltvTimeout(p.tx)
+              if (cltvTimeout > blockCount) {
+                log.info(s"delaying publication of txid=${p.tx.txid} until block=$cltvTimeout (current block=$blockCount)")
+                val cltvDelayedTxs1 = cltvDelayedTxs + (cltvTimeout -> (cltvDelayedTxs.getOrElse(cltvTimeout, Seq.empty) :+ p))
+                run(cltvDelayedTxs1, csvDelayedTx1)
+              } else {
+                publish(p)
+                run(cltvDelayedTxs, csvDelayedTx1)
+              }
+            } else {
+              log.info(s"some parent txs of txid=${p.tx.txid} are still unconfirmed (parent txids=${txWithRelativeDelay1.parentTxIds.mkString(",")})")
+              run(cltvDelayedTxs, csvDelayedTxs + (p.tx.txid -> txWithRelativeDelay1))
+            }
+          case None =>
+            log.warn(s"txid=${p.tx.txid} not found for parent txid=$parentTxId")
+            Behaviors.same
+        }
+
+      case WrappedCurrentBlockCount(blockCount) =>
+        val toPublish = cltvDelayedTxs.view.filterKeys(_ <= blockCount)
+        toPublish.values.flatten.foreach(tx => publish(tx))
+        run(cltvDelayedTxs -- toPublish.keys, csvDelayedTxs)
+
+      case PublishNextBlock(p) =>
+        val nextBlockCount = nodeParams.currentBlockHeight + 1
+        val cltvDelayedTxs1 = cltvDelayedTxs + (nextBlockCount -> (cltvDelayedTxs.getOrElse(nextBlockCount, Seq.empty) :+ p))
+        run(cltvDelayedTxs1, csvDelayedTxs)
+    }
+
+  private def publish(p: PublishTx): Future[ByteVector32] = {
+    p match {
+      case SignAndPublishTx(txInfo, commitments) =>
+        log.info("publishing tx: input={}:{} txid={} tx={}", txInfo.input.outPoint.txid, txInfo.input.outPoint.index, p.tx.txid, p.tx)
+        val publishF = txInfo match {
+          case tx: ClaimLocalAnchorOutputTx => publishLocalAnchorTx(tx, commitments)
+          case tx: HtlcTx => publishHtlcTx(tx, commitments)
+          case _ =>
+            log.error(s"ignoring unhandled transaction type ${txInfo.getClass.getSimpleName}")
+            Future.successful(ByteVector32.Zeroes)
+        }
+        publishF.recoverWith {
+          case t: Throwable if t.getMessage.contains("(code: -4)") || t.getMessage.contains("(code: -6)") =>
+            log.warn("not enough funds to publish tx, will retry next block: reason={} input={}:{} txid={}", t.getMessage, txInfo.input.outPoint.txid, txInfo.input.outPoint.index, p.tx.txid)
+            context.self ! PublishNextBlock(p)
+            Future.failed(t)
+          case t: Throwable =>
+            log.error("cannot publish tx: reason={} input={}:{} txid={}", t.getMessage, txInfo.input.outPoint.txid, txInfo.input.outPoint.index, p.tx.txid)
+            Future.failed(t)
+        }
+      case PublishRawTx(tx) =>
+        log.info("publishing tx: txid={} tx={}", tx.txid, tx)
+        publish(tx)
+    }
+  }
+
+  private def publish(tx: Transaction): Future[ByteVector32] = {
```

Comment:
```suggestion
  /**
   * This method uses a single thread to publish transactions so that it preserves the order of publication.
   * We need that to prevent concurrency issues while publishing parent and child transactions.
   */
  private def publish(tx: Transaction): Future[ByteVector32] = {
```

Reply:
Done in https://github.com/ACINQ/eclair/pull/1749/commits/515c2fa945eae1872d70af7c418e9ada6e521090